File: Pinball App 2/Pinball App 2/settings/SettingsManufacturerSupport.swift
import SwiftUI

enum ManufacturerBucket: String, CaseIterable, Identifiable {
    case modern
    case classic
    case other

    var id: String { rawValue }

    var label: String {
        switch self {
        case .modern:
            return "Modern"
        case .classic:
            return "Classic"
        case .other:
            return "Other"
        }
    }
}

extension Array where Element == PinballCatalogManufacturerOption {
    func filteredForBucket(_ bucket: ManufacturerBucket) -> [PinballCatalogManufacturerOption] {
        let classicTopIDs = self
            .filter { !$0.isModern }
            .sorted {
                if $0.gameCount != $1.gameCount { return $0.gameCount > $1.gameCount }
                return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
            .prefix(20)
            .map(\.id)
        let classicSet = Set(classicTopIDs)

        switch bucket {
        case .modern:
            return filter { $0.isModern }
        case .classic:
            return filter { classicSet.contains($0.id) }
                .sorted {
                    if $0.gameCount != $1.gameCount { return $0.gameCount > $1.gameCount }
                    return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
                }
        case .other:
            return filter { !$0.isModern && !classicSet.contains($0.id) }
        }
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/SettingsManufacturerSupport.kt
package com.pillyliu.pinprofandroid.settings

import com.pillyliu.pinprofandroid.library.CatalogManufacturerOption

internal enum class ManufacturerBucket(val label: String) {
    MODERN("Modern"),
    CLASSIC("Classic"),
    OTHER("Other"),
}

internal fun List<CatalogManufacturerOption>.filteredForBucket(
    bucket: ManufacturerBucket,
): List<CatalogManufacturerOption> {
    val classicTopIds = filter { !it.isModern }
        .sortedWith(compareByDescending<CatalogManufacturerOption> { it.gameCount }.thenBy { it.name.lowercase() })
        .take(20)
        .map { it.id }
        .toSet()
    return when (bucket) {
        ManufacturerBucket.MODERN -> filter { it.isModern }
        ManufacturerBucket.CLASSIC -> filter { it.id in classicTopIds }
            .sortedWith(compareByDescending<CatalogManufacturerOption> { it.gameCount }.thenBy { it.name.lowercase() })
        ManufacturerBucket.OTHER -> filter { !it.isModern && it.id !in classicTopIds }
    }
}
